dds/DCPS/DataWriterImpl_T.h

```diff
@@ -341,82 +341,40 @@ DataWriterImpl_T
 
   DDS::ReturnCode_t setup_serialization()
   {
-    /**
-     * TODO(iguessthislldo): Update when we can determine what encoding we need
-     * to use from from a data representation QoS.
-     */
-    encoding_mode_ = EncodingMode(
-      cdr_encapsulation() ?
-        Encoding::KIND_XCDR1 : Encoding::KIND_UNALIGNED_CDR,
-      swap_bytes());
-#if 0
-    /*
-     * We need to get the encoding kinds for encapsulated and non-encapsulated
-     * encoding modes. We do this by picking the first supported representation
-     * that is supported by the mode, which is sorta what the XTypes spec says
-     * to do, except it only assumes one mode. One mode (but not both) may be
-     * left set to KIND_UNKNOWN, disabling it and preventing it from matching
-     * with readers just available through that mode.
-     */
-    Encoding::Kind encap_kind = Encoding::KIND_UNKNOWN;
-    Encoding::Kind nonencap_kind = Encoding::KIND_UNKNOWN;
-    DDS::DataRepresentationIdSeq repIds =
-      get_effective_data_rep_qos(qos_.representation.value);
-    for (CORBA::ULong i = 0; i < repIds.length(); ++i) {
-      const DDS::DataRepresentationId_t repr = repIds[i];
-      const Encoding::Kind kind =
-        repr_ext_to_encoding_kind(repr, MarshalTraitsType::extensibility());
-      switch (kind) {
-      case Encoding::KIND_CDR_UNALIGNED:
-        if (nonencap_kind == Encoding::KIND_UNKNOWN) {
-          nonencap_kind = kind;
+    DDS::DataRepresentationIdSeq repIds = get_effective_data_rep_qos(qos_.representation.value);
+    bool success = false;
+    Encoding::Kind encoding_kind;
+    if (cdr_encapsulation()) {
+      for (CORBA::ULong i = 0; i < repIds.length(); ++i) {
+        // TODO(sonndinh): Do we need to return failure when encountering an invalid representation
+        if (repr_to_encoding_kind(repIds[i], encoding_kind) &&
+            (Encoding::KIND_XCDR1 == encoding_kind ||
+             Encoding::KIND_XCDR2 == encoding_kind)) {
+          encoding_mode_ = EncodingMode(encoding_kind, swap_bytes());
```

Comment:
Maybe set `encoding_kind` instead and move this down below the `success` check, so it's shared with the non-encapsulated case?

Reply:
Actually `repr_to_encoding_kind` is setting the encoding kind, so just break.

Reply:
Per your review above for the non-encapsulated case, the for-loop is removed and there is no duplicate code now.